core/trino-spi/src/main/java/io/trino/spi/predicate/SortedRangeSet.java

```diff
@@ -553,6 +568,128 @@ public SortedRangeSet intersect(ValueSet other)
+                else {
+                    // The intersectionEndIndex is an exclusive index that needs to be increased when an overlapping RangeSet was found
+                    intersectionEndIndex++;
+                }
+            }
+            // test if testRange covers whole probeSet
+            if (insertionStartIndex == 0 && intersectionEndIndex >= probeEnd) {
+                RangeView startRange = probeRangeSet.getRangeView(0);
+                RangeView endRange = probeRangeSet.getRangeView(probeEnd - 1);
+
+                Optional<RangeView> startRangeIntersection = startRange.tryIntersect(current);
+                Optional<RangeView> endRangeIntersection = endRange.tryIntersect(current);
+                if (startRangeIntersection.isPresent() && endRangeIntersection.isPresent() &&
+                        startRangeIntersection.get().compareTo(startRange) == 0 && endRangeIntersection.get().compareTo(endRange) == 0) {
+                    return probeRangeSet;
+                }
+            }
+            if (blockBuilder == null) {
+                blockBuilder = type.createBlockBuilder(null, 2 * (testEnd + probeEnd));
+                inclusive = new boolean[2 * (testEnd + probeEnd)];
+            }
+            int probeIndex = insertionStartIndex;
+            while (probeIndex < intersectionEndIndex) {
+                RangeView probeRange = probeRangeSet.getRangeView(probeIndex);
+                // intersection at edges as [1, 9], [12, 18] intersected with [7, 15], [17, 21] should end up as [7, 9], [12, 15], [17, 18]
+                if (probeIndex == insertionStartIndex || probeIndex + 1 >= intersectionEndIndex) {
+                    Optional<RangeView> intersect = probeRange.tryIntersect(current);
+                    if (intersect.isPresent()) {
+                        writeRange(type, blockBuilder, inclusive, resultIndex, intersect.get());
+                        resultIndex++;
+                    }
+                    probeIndex++;
+                }
+                else {
+                    Block block = probeRangeSet.getSortedRanges();
+                    if (block instanceof DictionaryBlock || block instanceof ValueBlock) {
+                        int size = intersectionEndIndex - probeIndex - 1;
+                        int offset = probeIndex * 2;
+                        if (block instanceof DictionaryBlock) {
+                            copyDictionaryBlock(blockBuilder, inclusive, probeRangeSet, offset, resultIndex, size);
+                        }
+                        else {
+                            copyValueBlock(blockBuilder, inclusive, probeRangeSet, offset, resultIndex, size);
+                        }
+                        probeIndex += size;
+                        resultIndex += size;
+                    }
+                    else {
+                        // RLE
+                        writeRange(type, blockBuilder, inclusive, resultIndex, probeRange);
```

Comment:
This code is assuming you have an RLE, but technically this could be a LazyBlock.  I don't beleive the current code allows for this, but it would be better to explicitly test that is is RLE.  You can use a moder switch to simplify this since the Block class is sealed.